crates/pipeline-manager/src/api/main.rs

```diff
@@ -59,14 +59,113 @@ out of the SQL program, but also has accompanying metadata and configuration par
+
+In case of a response, usually it is the Feldera endpoint that generated it:
+- If it is success (2xx), it will return whichever body belongs to the success response.
+- Otherwise, if it is an error (4xx, 5xx), it will return a Feldera error response JSON body
+  which will have an application-level `error_code`.
+
+However, there are two notable exceptions when the response is not generated by the Feldera
+endpoint:
+- If the HTTP server, to which the endpoint belongs, encountered an issue, it might return
+  4xx (e.g., for an unknown endpoint) or 5xx error codes by itself (e.g., when it is initializing).
+- If it is behind a (reverse) proxy, the proxy can return error codes by itself,
+  for example BAD GATEWAY (502) or GATEWAY TIMEOUT (504).
+
+As such, it is not guaranteed that the (4xx, 5xx) will have a Feldera error response JSON body
+in these latter cases.
+
+### Error handling and retrying
+
+The error type returned by the client should distinguish between the error responses generated
+by Feldera endpoints themselves (which have a Feldera error response body) and those that are
+generated by other sources.
+
+In order for a client operation (e.g., `pipeline.resume()`) to be robust (i.e., not fail due to
+a single HTTP request not succeeding) the client should use a retry mechanism if the operation
+is idempotent. The retry mechanism must however have a time limit, after which it times out.
+This guarantees that the client operation is eventually responsive, which enables the script
+it is a part of to not hang indefinitely on Feldera operations and instead be able to decide
+by itself whether and how to proceed. If no response is returned, the mechanism should generally
+retry. When a response is returned, the decision whether to retry can generally depend on the status
+code: especially the status codes 502, 503 and 504 should be considered as transient errors.
+Finer grained retry decisions should be made by taking into account the application-level
+`error_code` if the response body was indeed a Feldera error response body.
+
+## Feldera client errors (4xx)
+
+_Client behavior:_ clients should immediately return with an error when they get back a 4xx status
+code, as it means the request will very likely not succeed even if it is sent again.
+
+- **BAD REQUEST (400)**: invalid user request (general).
+  - _Example:_ the new pipeline name `example1@~` contains invalid characters.
+
+- **UNAUTHORIZED (401)**: the user is not authorized to issue the request.
+  - _Example:_ an invalid API key is provided.
+
+- **NOT FOUND (404)**: a resource required to exist in order to process the request was not found.
+  - _Example:_ a pipeline named `example` does not exist when trying to update it.
+
+- **CONFLICT (409)**: there is a conflict between the request and a relevant resource.
+  - _Example:_ a pipeline named `example` already exists.
+  - _Example:_ another transaction is already in process.
```

Comment:
here I asked last time: why should this not be retried? Maybe the other transaction will finish in the meantime.

Reply:
Changed client behavior to:

_Client behavior:_ clients should generally return with an error when they get back a 4xx status
code, as it usually means the request will likely not succeed even if it is sent again. Certain
errors for some types of requests might make use of a timed retry mechanism, for instance a
transaction already being in progress leading to a temporary conflict.